### PR TITLE
fix gcs_placement_group_test flaky

### PR DESCRIPTION
**Pull request by larrylian**
## Why are these changes needed?
This test case is very old, and the logic is very complicated and not particularly reasonable. I simply refactored now. However, you can try to refactor the entire test case of gcs_placement_group_test in the future.

## Related issue number

[Test] gcs_placement_group_manager_test flaky in the master #30573

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Hmm I am confused about the reasoning. The PR broke this test that never broke before. Can you tell us more detail?

**Comment by larrylian** [comment]
> Hmm I am confused about the reasoning (`very complicated and not particularly reasonable`). The PR broke this test that never broke before. Can you tell us more detail? The fact that this test never broke before means there was some logic issue in the merged PR.

After the scheduling is completed, my PR[#24875](https://github.com/ray-project/ray/pull/24875) will check whether the PG still has unplaced bundles. If so, the scheduling will continue. 

This feature causes this use case to fail now. So this test case needs to be refactored. 
 In fact, this test case is not particularly correct, so don't worry too much about the specific cause of the error.

**Comment by rkooo567** [comment]
@larrylian I need to understand the change to review the code. Can you at least tell me the high level behavior of tests before -> after refactoring? That'd be really helpful to review the PR!  Thank you!

**Comment by larrylian** [comment]
> @larrylian Got it. But I still need to understand the change to review the code. Can you at least tell me the high level behavior of tests before -> after refactoring? That'd be really helpful to review the PR! Thank you!

OK. @rkooo567 
TestRescheduleWhenNodeDead This test case is to test the scenario of rescheduling PG after  node dies. 
So use case behavior. 
1. PG scheduling success -> NodeDead —> PG rescheduling success. 
2. Then NodeDead->PG rescheduling failed.

Its original writing was not rigorous enough. PG executes NodeDead before scheduling sucesss. So it will cause many small problems.
Also, it originally verified that the rescheduled PG will be scheduled first. But now we don't have this feature anymore. So delete this point.

**Comment by rkooo567** [comment]
> Its original writing was not rigorous enough. PG executes NodeDead before scheduling sucesss. So it will cause many small problems.

But `PG executes NodeDead before scheduling sucesss` is a possible scenario right? We should still test this case? 

> Also, it originally verified that the rescheduled PG will be scheduled first. But now we don't have this feature anymore. So delete this point.

I don't think it is true. What do you mean by this is not supported anymore? It is important to avoid deadlocking. (rescheduled ones should be higher priority than regular pg scheduling)

File: src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc [hunk 1]
```diff
@@ -431,43 +431,40 @@ TEST_F(GcsPlacementGroupManagerTest, TestRescheduleWhenNodeDead) {
...
+  OnPlacementGroupCreationSuccess(placement_group);
+
   // If a node dies, we will set the bundles above it to be unplaced and reschedule the
-  // placement group. The placement group state is set to `RESCHEDULING` and will be
-  // scheduled first.
+  // placement group. The placement group state is set to `RESCHEDULING`
   mock_placement_group_scheduler_->group_on_dead_node_ =
       placement_group->GetPlacementGroupID();
   mock_placement_group_scheduler_->bundles_on_dead_node_.push_back(0);
   gcs_placement_group_manager_->OnNodeDead(NodeID::FromRandom());
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:445]: This part kind of doesn't make sense. How is killing a random node cause rescheduling? Isn't it supposed to be a node that pg is scheduled on?

**Review by rkooo567 — COMMENTED** [review]
I agree the current test is weird, but I think the fix will completely change the test (and only test the case 1 below iiuc). We should basically test 3 things.

1. Regular case (success -> node fail -> rescheduling)
2. Failure before success (not succeeded yet -> node fail -> rescheduling)
3. When 1 or 2 happens, the rescheduling pg should have the higher priority than regular pg scheduling

Why don't we just rewrite tests in this way?

**Comment by larrylian** [comment]
> > Its original writing was not rigorous enough. PG executes NodeDead before scheduling sucesss. So it will cause many small problems.
> 
> But `PG executes NodeDead before scheduling sucesss` is a possible scenario right (in real world, node can fail before pg succeeds to be scheduled)? We should still test this case? Can you add a test case to verify this? (Rescheduling happens before pg succeeds to be created)
> 
> > Also, it originally verified that the rescheduled PG will be scheduled first. But now we don't have this feature anymore. So delete this point.
> 
> What do you mean by this is not supported anymore? You mean this feature was broken? It is an important feature to avoid deadlocking. (RESCHEDULING pgs should be scheduled before PENDING pgs)


What I mean is that now when PG reschedules when a node fails, it will no longer be forced to join the first one in the pendingQueue. Now the backoff mechanism has been added to the PendingQueue, so there is no need to worry about deadlocking.

File: src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc (hunk 1 above)
Comment (larrylian) [reply src/ray/gcs/gcs_server/test/gcs_placement_group_manager_test.cc:445]: This is actually where the original code was not written well enough. Its mock is written like this, and it is difficult for me to explain it. I think this test case is simply fixed for now. Then it is better to refactor this gcs_placement_group_manager_test in a new way in the future.

**Comment by larrylian** [comment]
> I agree the current test is weird, but I think the fix will completely change the test (and only test the case 1 below iiuc). We should basically test 3 things.
> 
> 1. Regular case (success -> node fail -> rescheduling)
> 2. Failure before success (not succeeded yet -> node fail -> rescheduling)
> 3. When 1 or 2 happens, the rescheduling pg should have the higher priority than regular pg scheduling
> 
> Why don't we just rewrite tests in this way?

1.  I agree with the 1 point. That's what I'm testing now.
2. The 2 point is that an exception occurs during the test scheduling process. There is no way to test it here, there should be tests elsewhere.
3. The 3 point is not necessary according to the current PendingQueue mechanism. Just get scheduled as early as possible. So there is no need to test this.

**Comment by rkooo567** [comment]
Let me just try writing a test on this PR. 3 is critical for the fault tolerance protocol. Imagine this scenario; 

pg A CREATED
node A {CPU: 1}, node B {CPU: 1}

node B is dead

pg B PENDING
{CPU: 1}

if rescheduling is not prioritized, pg A will stay partially created, which is the violence of the atomicity. If your PR changed this behavior, we should revert that PR

**Comment by larrylian** [comment]
> Let me just try writing a test on this PR. 3 is critical for the fault tolerance protocol. Imagine this scenario;
> 
> pg A CREATED node A {CPU: 1}, node B {CPU: 1}
> 
> node B is dead
> 
> pg B PENDING {CPU: 1}
> 
> if rescheduling is not prioritized, pg A will stay partially created, which is the violence of the atomicity. If your PR changed this behavior, we should revert that PR

My PR did not change the OnNodeDead code. So there is no change to this behavior.

**Review by rkooo567 — APPROVED** [review]
Since I agree the original test was bad, let me merge it and unblock release tests first. I will try adding 2 other cases I mentioned from the comment from my end!

**Pull request merged**